proposals/0046-optimistic-cluster-restart-automation.md

```diff
@@ -0,0 +1,321 @@
+remove some false positive cases.
+
+However, it's also overkill to repair every block presented by others. When
+`LastVotedForkSlots` messages are being received and aggregated, a validator
+can categorize blocks missing locally into 3 categories: ignored, must-have, 
+and unsure. Depending on the stakes of validators currently in restart, some
+slots with too few stake can be safely ignored, some have enough stake they
+should definitely be repaired, and the rest would be undecided pending more
+confirmations.
+
+Assume `RESTART_STAKE_THRESHOLD` is 80% and that 5% restarted validators can
+change their votes from what they voted before the restart due to mistakes or 
+malicious behavior.
+
+When only 5% validators are in restart, everything is in "unsure" category.
+
+When 67% validators are in restart, any slot with less than
+67% - 5% - (100-67%) = 29% is in "ignored" category, because even if all 
+validators join the restart, the slot will not get 67% stake. When this 
+threshold is less than 33%, we temporarily put all blocks with >33% stake into  
+"must-have" category to speed up repairing. Any slot with between 29% and 33% 
+stake is "unsure".
+
+When 80% validators are in restart, any slot with less than
+67% - 5% - (100-80%) = 42% is in "ignored" category, the rest is "must-have".
+
+From above examples, we can see the "must-have" threshold changes dynamically 
+depending on how many validators are in restart. The main benefit is that a
+block will only move from "must-have/unsure" to "ignored" as more validators 
+join the restart, not vice versa. So the list of blocks a validator needs to
+repair will never grow bigger when more validators join the restart.
+
+Once the validator gets LastVotedForkSlots, it can draw a line which are the
+"must-have" blocks. When all the "must-have" blocks are repaired and replayed,
+it can proceed to step 3.
+
+### 3. "silent repair phase": Gossip current heaviest fork
+
+The main goal of this step is to "vote" the heaviest fork to restart from.
+
+We use a new Gossip message `HeaviestFork`, its fields are:
+
+* `slot`: `u64` slot of the picked block.
+* `hash`: `Hash` bank hash of the picked block.
+* `received`: `u8` total percentage of stakes of the validators it received
+`HeaviestFork` messages from.
+
+After receiving `LastVotedForkSlots` from the validators holding stake more 
+than  `RESTART_STAKE_THRESHOLD` and repairing slots in "must-have" category,
+replay all blocks and pick the heaviest fork as follows:
+
+1. For all blocks with more than 67% votes, they must be on picked fork.
+
+2. If a picked block has more than one children, check if the votes on the
+heaviest child is over threshold:
+
+   1. If vote_on_child + stake_on_validators_not_in_restart >= 62%, pick child.
+For example, if 80% validators are in restart, child has 42% votes, then
+42 + (100-80) = 62%, pick child. 62% is chosen instead of 67% because 5%
+could make the wrong votes.
```

Comment:
```suggestion
Define the "must have" threshold to be 62%. If a picked block has more than one child, we check for each child `vote_on_child + stake_on_validators_not_in_restart >= 62%`. If so, pick the child.
For example, if 80% validators are in restart, child has 42% votes, then
42 + (100-80) = 62%, pick child. 62% is chosen instead of 67% because 5%
could make the wrong votes.
```

Reply:
Done.